aptos-move/aptos-code-analyzer/src/main.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_code_analyzer::{
+    passes::{
+        analyze_modules::ModuleAnalysis,
+        collect_cost_cfgs::{CollectCostCFGs, Counter, Extension},
+    },
+    ModulePass,
+};
+use aptos_language_e2e_tests::executor::FakeExecutor;
+use clap::{Parser, Subcommand};
+use move_binary_format::CompiledModule;
+use std::{
+    fs,
+    fs::File,
+    io::{BufRead, BufReader},
+    path::PathBuf,
+};
+
+#[derive(Subcommand)]
+pub enum Cmd {
+    AnalyzeCode {
+        output_dir: PathBuf,
+    },
+    BuildCfgs {
+        output_dir: PathBuf,
+        #[clap(value_enum)]
+        output_extension: Extension,
+        #[clap(value_enum)]
+        counter: Counter,
+    },
+}
+
+#[derive(Parser)]
+pub struct Argument {
+    // Path to a file which contains all paths to bytecode modules.
+    state_view_path: PathBuf,
+    #[clap(subcommand)]
+    cmd: Cmd,
+}
+
+fn main() {
+    let args = Argument::parse();
+
+    match args.cmd {
+        Cmd::AnalyzeCode { output_dir } => {
+            let (_, modules) = initialize(args.state_view_path);
+            let mut pass = ModuleAnalysis::new(output_dir);
+            for module in modules.into_iter() {
+                pass.run_on_module(&module)
+            }
+            pass.finish();
+        },
+        Cmd::BuildCfgs {
+            output_dir,
+            output_extension,
+            counter,
+        } => {
+            let (exec, modules) = initialize(args.state_view_path);
+            let mut pass = CollectCostCFGs::new(
+                &output_dir,
+                &output_extension,
+                exec.get_state_view(),
+                &counter,
+            );
+            for module in modules.into_iter() {
+                pass.run_on_module(&module)
+            }
+        },
+    }
+}
+
+fn initialize(state_view_path: PathBuf) -> (FakeExecutor, Vec<CompiledModule>) {
+    let state_view_file = File::open(state_view_path.as_path())
+        .expect("Must provide a file with paths to all bytecode modules");
+    let reader = BufReader::new(state_view_file);
+
+    // State view doesn't have any modules by default.
+    let mut exec = FakeExecutor::from_head_genesis();
+    let mut modules = Vec::new();
+
+    let mut success_cnt = 0;
+    let mut total_cnt = 0;
+    for line in reader.lines() {
+        let line = line.unwrap();
+        let mut iter = line.split_ascii_whitespace();
+        let bytecode_path = iter.next().unwrap();
+
+        let bytes = fs::read(bytecode_path).expect("Should be able to read module data");
+        match CompiledModule::deserialize(&bytes) {
+            Ok(module) => {
```

Comment:
I'm really worried that we do not use the canonical way to analyze Move code which is the Move model. Is there something here which you supposedly can only do with the raw bytecode instead of the stackless bytecode?

We may not have an easy way today to connect the Move model if you do not have source, though @rahxephon89 has worked this out previously.


We can move forward with this if it blocks something, but we should not further extend this. We don't want to build up a 2nd code analysis framework for Move.

Reply:
There is no rush to land this, I am just pushing code I use myself to gather statistics and thought it would be nice if others can reuse it. We should definitely find the best way to merge this with existing tools, 100% agree.

I had a look at the Move model, and there are a few problems for my use case. Move model is great for compiler, and for analyses passes (and should be reused). What I am trying to do is to analyze the bytecode on chain, that is get gas costs, number of actual bytecode instructions, global accesses. If I use Move model abstraction, I lose this precise information? 

I guess it's all about different levels of abstractions: for these use cases compiled module is what is needed, but for things which are more high level and use data flow analysis for example, we should use a move model.

Reply:
You can access the bytecode directly. Currently there is really an example missing how to setup the plumbing and probably also a few code changes needed for this use case, we can look later on this.